tripy/examples/diffusion/model.py

```diff
@@ -0,0 +1,775 @@
+# https://arxiv.org/pdf/2112.10752.pdf
+# https://github.com/ekagra-ranjan/huggingface-blog/blob/main/stable_diffusion.md
+# Adapted from https://github.com/tinygrad/tinygrad/blob/master/examples/stable_diffusion.py
+
+import gzip, math, re, os, pathlib, platform, tempfile, hashlib
+import urllib.request
+from functools import lru_cache, reduce
+from tqdm import tqdm
+from collections import namedtuple
+from typing import List, Callable, Optional, Union
+
+import numpy as np
+import tripy as tp
+
+# convenience methods adapted from tinygrad/tensor.py (https://docs.tinygrad.org/tensor/ops/)
+def scaled_dot_product_attention(
+    query: tp.Tensor,
+    key: tp.Tensor,
+    value: tp.Tensor,
+    embedding_dim: Optional[int] = None,
+    attn_mask: Optional[tp.Tensor] = None,
+    is_causal: bool = False,
+) -> tp.Tensor:
+    """
+    Computes scaled dot-product attention.
+    `self` is the query tensor, `key` is the key tensor, and `value` is the value tensor.
+
+    - Described: https://paperswithcode.com/method/scaled
+    - Paper: https://arxiv.org/abs/1706.03762v7
+    """
+
+    if is_causal:  # this path is not called in demoDiffusion
+        target_shape = query.shape[-2:-1] + key.shape[-2:-1]
+        # TODO: #228: WAR to prevent computing output rank in infer_rank for reshape
+        target_shape.trace_tensor.shape = (2,)
+        attn_mask = tp.cast(tp.tril(tp.ones(target_shape)), tp.bool)
+    if attn_mask is not None and attn_mask.dtype == tp.bool:
+        attn_mask = tp.where((attn_mask == 0), tp.ones_like(attn_mask) * -float("inf"), tp.zeros_like(attn_mask))
+    qk = query @ tp.transpose(key, -2, -1) / math.sqrt(embedding_dim)
+    return tp.cast(tp.softmax((qk + attn_mask) if attn_mask is not None else qk, -1), query.dtype) @ value
+
+
+def sequential(input: tp.Tensor, ll: List[Callable[[tp.Tensor], tp.Tensor]]):
+    """
+    Applies a sequence of functions to `self` chaining the output of each function to the input of the next.
+    """
+    return reduce(lambda x, f: f(x), ll, input)
+
+
+# convenience for dynamic reshapes
+one_shape = tp.Shape(tp.ones((1,), dtype=tp.int32))
+
+# TODO: change to linear layers?
+class AttnBlock(tp.Module):
+    def __init__(self, in_channels):
+        self.group_norm = tp.GroupNorm(32, in_channels)
+        self.to_q = tp.Linear(in_channels, in_channels)
+        self.to_k = tp.Linear(in_channels, in_channels)
+        self.to_v = tp.Linear(in_channels, in_channels)
+        self.to_out = [tp.Linear(in_channels, in_channels)]
+        self.in_channels = in_channels
+
+    # adapted from AttnBlock in ldm repo
+    def __call__(self, x):
+        h_ = self.group_norm(x)
+
+        b, c, h, w = h_.shape[0], h_.shape[1], h_.shape[2], h_.shape[3]
```

Comment:
Since we can determine the length of the shape after @slyubomirsky's PR, could we use `__iter__` to do the unpacking? Then this would be:
```py
b, c, h, w = h_.shape
```